core/ArchiveProcessor/Loader.php

```diff
@@ -389,12 +390,63 @@ public function canSkipThisArchive()
         $hasSiteVisitsBetweenTimeframe = $this->hasSiteVisitsBetweenTimeframe($idSite, $params->getPeriod());
         $hasChildArchivesInPeriod = $this->dataAccessModel->hasChildArchivesInPeriod($idSite, $params->getPeriod());
 
+        if ($this->canSkipArchiveForSegment()) {
+            return true;
+        }
+
         return $isWebsiteUsingTracker
             && !$isArchivingForcedWhenNoVisits
             && !$hasSiteVisitsBetweenTimeframe
             && !$hasChildArchivesInPeriod;
     }
 
+    public function canSkipArchiveForSegment()
+    {
+        $params = $this->params;
+
+        if ($params->getSegment()->isEmpty()) {
+            return false;
+        }
+
+        $idSite = $params->getSite()->getId();
+        $periodEnd = $params->getPeriod()->getDateEnd();
+        $segmentHash = $params->getSegment()->getHash();
+        /** @var SegmentArchiving */
+        $segmentArchiving = StaticContainer::get(SegmentArchiving::class);
+        $segmentInfo = $segmentArchiving->findSegmentForHash($segmentHash, $idSite);
+
+        if (!$segmentInfo) {
+            return false;
+        }
+
+        $segmentArchiveStartDate = $segmentArchiving->getReArchiveSegmentStartDate($segmentInfo);
+
+        if ($segmentArchiveStartDate !==null && $segmentArchiveStartDate->isLater($periodEnd->getEndOfDay())) {
+            $doneFlag = Rules::getDoneStringFlagFor(
+                [$idSite],
+                $params->getSegment(),
+                $params->getPeriod()->getLabel(),
+                $params->getRequestedPlugin()
+            );
+
+            // no specific report request
+            // if there is no invalidation where the report is null, we can skip
+            // if we have invalidations for the period and name, but only for a specific reports, we can skip
+            if (!$params->getArchiveOnlyReport() && !$this->dataAccessModel->hasInvalidationForPeriodAndName($idSite, $params->getPeriod(), $doneFlag)) {
+                return true;
+            }
+
+            // specific report request
+            // we only want to rearchive if we have invalidation for that report
+            // if we don't find invalidation for that report, we can skip
+            if ($params->getArchiveOnlyReport() && !$this->dataAccessModel->hasInvalidationForPeriodAndName($idSite, $params->getPeriod(), $doneFlag, $params->getArchiveOnlyReport())) {
+                return true;
+            }
```

Comment:
Could the two `if`s above be simplified to just:

```
return !$this->dataAccessModel->hasInvalidationForPeriodAndName($idSite, $params->getPeriod(), $doneFlag, $params->getArchiveOnlyReport());
```
?

Since archive only report will be empty when it's not set, we can just pass it as is, right?

Reply:
good question. To be honest, I'm not sure, can't remember exactly, but probably you are right. Needs to be tested to make sure.

Reply:
I thought and walked through it, it should work the way you recommended, I updated the PR.